interfaces/devdoc/malloc_multi_flex_requirements.md

```diff
@@ -0,0 +1,124 @@
+# `MALLOC_MULTI_FLEX` requirements
+
+## Overview
+
+`MALLOC_MULTI_FLEX` allows to allocate memory for a structure containing multiple variable sized array members using a single malloc.
+
+## Exposed API
+
+```c
+#define DEFINE_MALLOC_MULTI_FLEX(type, ...)
+    ...
+
+#define MALLOC_MULTI_FLEX(type)
+    ...
+```
+
+## Example - 
+
+Consider there are two structs as defined below -
+
+```c
+typedef struct INNER_STURCT_TAG
+{
+    uint32_t inner_int_1;
+    uint64_t inner_int_2;
+}INNER_STRUCT;
+
+typedef struct PARENT_STRUCT_TAG
+{
+    uint64_t int_1;
+    uint32_t* array_1;
+    uint32_t int_2;
+    uint64_t* array_2;
+    INNER_STRUCT* array_3;
+    uint32_t int_3;
+}PARENT_STRUCT;
+```
+
+Lets say the user wants to allocate memory for `PARENT_STRUCT` which has three array members:
+- array_1 of type `uint32_t`
+- array_2 of type `uint64_t`
+- array_3 of type `INNER_STRUCT`
+
+The caller first needs to specify the struct and member details using `DEFINE_MALLOC_MULTI_FLEX` macro in `.c` file to create a custom `malloc_multi_flex_<type>` function as defined below -
+
+```c
+DEFINE_MALLOC_MULTI_FLEX(PARENT_STRUCT, array_1, array_2, array_3)
+```
+
+which will get expanded as following -
+
+```c
+static void* malloc_multi_flex_PARENT_STRUCT(size_t parent_struct_size, uint32_t array_1_count, uint32_t array_1_size, uint32_t array_2_count, uint32_t array_2_size, uint32_t array_3_count, uint32_t array_3_size) 
+{
+    // perform overflow checks while calculating total size to be allocated
+    size_t size_tracker = parent_struct_size; 
+    if ((array_1_size != 0 && SIZE_MAX / array_1_size < array_1_count) || (SIZE_MAX - size_tracker < array_1_count * array_1_size)) 
+    {
+        return NULL;
+    } 
+    size_tracker += array_1_count * array_1_size;
```

Comment:
>ize_tracker += array_1_count * array_1_size; [](http://example.com/codeflow?start=5&length=44)

I think one more thing we need to take into consideration as we do something more complex like this is alignment.

Possibly helpful reading [https://en.cppreference.com/w/c/language/object#Alignment](https://en.cppreference.com/w/c/language/object#Alignment)

I'm fairly sure this will work if we also compute the `alignof` for the type of the array element and add padding before each array to make sure it is aligned.

But to be honest, this isn't something I'm completely familiar with and I'd like to defer to maybe Dan or Andrei to review this as well. Andrei has an example of this in `async_op_create`

Reply:
+1. When I read the synopsis of this I was thinking that we have to take into account alignment. So I was then looking for this specific part where we compute the size ... 

I believe we should be using alignof